marker_rustc_driver/src/lib.rs

```diff
@@ -0,0 +1,388 @@
+
+use crate::conversion::rustc::RustcConverter;
+
+// region replace rust toolchain dev
+const RUSTC_TOOLCHAIN_VERSION: &str = "nightly-2023-10-05";
+// endregion replace rust toolchain dev
+
+struct DefaultCallbacks {
+    env_vars: Vec<(&'static str, String)>,
+}
+impl rustc_driver::Callbacks for DefaultCallbacks {
+    fn config(&mut self, config: &mut rustc_interface::Config) {
+        let env_vars = std::mem::take(&mut self.env_vars);
+        config.parse_sess_created = Some(Box::new(move |sess| {
+            register_tracked_env(sess, &env_vars);
+        }));
+    }
+}
+
+struct MarkerCallback {
+    env_vars: Vec<(&'static str, String)>,
+    lint_crates: Vec<LintCrateInfo>,
+}
+
+impl rustc_driver::Callbacks for MarkerCallback {
+    fn config(&mut self, config: &mut rustc_interface::Config) {
+        let env_vars = std::mem::take(&mut self.env_vars);
+        let lint_crates = self.lint_crates.clone();
+        config.parse_sess_created = Some(Box::new(move |sess| {
+            register_tracked_env(sess, &env_vars);
+            register_tracked_files(sess, &lint_crates);
+        }));
+
+        // Clippy explicitly calls any previous `register_lints` functions. This
+        // will not be done here to keep it simple and to ensure that only known
+        // code is executed.
+        assert!(config.register_lints.is_none());
+        let lint_crates = std::mem::take(&mut self.lint_crates);
+
+        config.register_lints = Some(Box::new(move |_sess, lint_store| {
+            // It looks like it can happen, that the `config` function is called
+            // with a different thread than the actual lint pass later, how interesting.
+            // This will not make sure that the adapter is always initiated.
+            if let Err(err) = lint_pass::RustcLintPass::init_adapter(&lint_crates) {
+                err.print();
+                // FIXME: we need to figure out the way to run the initialization code
+                // earlier where we can cleanly report an error and exit the process.
+                std::process::exit(1);
+            }
+            // Register lints from lint crates. This is required to have rustc track
+            // the lint level correctly.
+            let lints: Vec<_> = lint_pass::RustcLintPass::marker_lints()
+                .into_iter()
+                .map(RustcConverter::static_to_lint)
+                .collect();
+
+            lint_store.register_lints(&lints);
+
+            lint_store.register_late_pass(|_| Box::new(lint_pass::RustcLintPass));
+        }));
+    }
+}
+
+fn register_tracked_env(sess: &mut rustc_session::parse::ParseSess, vars: &[(&'static str, String)]) {
+    use rustc_span::Symbol;
+    let env = sess.env_depinfo.get_mut();
+
+    for (key, value) in vars {
+        env.insert((Symbol::intern(key), Some(Symbol::intern(value))));
+    }
+}
+
+fn register_tracked_files(sess: &mut rustc_session::parse::ParseSess, lint_crates: &[LintCrateInfo]) {
```

Comment:
Rustc uses caching where possible. This function, and the function above, register relevant files and environment values which should be checked, when rustc decides if a cache can be reused or not.

This one mainly registers the `Cargo.toml` file, as that's where Marker's lints are configured and the lint crate files. It basically just ensures that we don't use caches if any lint crate was updated. Do you maybe want to make this one public as well? I'm not sure how dylint handles this kind of thing.

This is just a suggestion, if you handle this differently or don't need it, I'm happy to merge this as is :)

Reply:
> Do you maybe want to make this one public as well?

It's kind of you to ask, but I don't think it's necessary.

If "`marker_rustc_driver` as a library" feels "incomplete" to you without this function, I can sympathize with that feeling. But I would recommend to keep the interface as small as possible until me or someone else complains. :)

Reply:
I just wanted to mention it, since i[#294]t was a thing we ran into after some testing :)

Reply:
I'm curious now (because caching has been the source of several bugs in Dylint). Is there a direct connection between caching and #294?

Reply:
The reference to #294 seems to be an accidental copy-pasta :spaghetti:. Sorry, about that.

There is a direct connection between the function I marked here and caching. This is the PR that prevented caching for Marker: https://github.com/rust-marker/marker/pull/211

Do you have some examples for caching related bugs in Dylint and how you solved them?

Reply:
> Do you have some examples for caching related bugs in Dylint and how you solved them?

Two that come to mind are:
- https://github.com/trailofbits/dylint/issues/650
  - fixed by: https://github.com/trailofbits/dylint/pull/813
- https://github.com/trailofbits/dylint/issues/856
  - fixed by: https://github.com/trailofbits/dylint/pull/866